src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -845,25 +850,25 @@ int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data(bufferlist& bl
             filter->get_policy_driver()->get_cache_policy()->update(dpp, oid, ofs, bl.length(), version, *y);
 
 	    /* Store block in directory */
-            if (!blockDir->exist_key(dpp, &block, *y)) {
-              if (blockDir->set(dpp, &block, *y) < 0) //should we revert previous steps if this step fails?
-		ldpp_dout(dpp, 0) << "D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::" << __func__ << "(): BlockDirectory set method failed." << dendl;
-            } else {
-              existing_block.blockID = block.blockID;
-              existing_block.size = block.size;
-              if (blockDir->get(dpp, &existing_block, *y) < 0) {
-                ldpp_dout(dpp, 0) << "Failed to fetch existing block for: " << existing_block.cacheObj.objName << " blockID: " << existing_block.blockID << " block size: " << existing_block.size << dendl;
-              } else {
-                if (existing_block.version != block.version) {
-                  if (blockDir->del(dpp, &existing_block, *y) < 0) //delete existing block
-                    ldpp_dout(dpp, 0) << "D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::" << __func__ << "(): BlockDirectory del method failed." << dendl;
-                  if (blockDir->set(dpp, &block, *y) < 0) //new versioned block will have new version, hostsList etc, how about globalWeight?
-                    ldpp_dout(dpp, 0) << "D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::" << __func__ << "(): BlockDirectory set method failed." << dendl;
-                } else {
-                if (blockDir->update_field(dpp, &block, "blockHosts", dpp->get_cct()->_conf->rgw_local_cache_address, *y) < 0)
-                  ldpp_dout(dpp, 0) << "D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::" << __func__ << "(): BlockDirectory update_field method failed for hostsList." << dendl;
-                }
-              }
+	    existing_block.blockID = block.blockID;
+	    existing_block.size = block.size;
+
+	    if ((ret = blockDir->get(dpp, &existing_block, *y)) == -ENOENT) {
```

Comment:
Trying to re-organize the code a bit, eliminating check for version, just overwrite the new latest version, how about the following:
if ((ret = blockDir->get(dpp, &existing_block, *y)) == 0) { //block exists
  //check if local host is NOT part of block.hostList, then add localhost
  //copy over other fields from existing_block to block (excluding version)
} else if (ret == -ENOENT) //block doesn't exist {
  //add localhost to block.hostList
}
if (blockDir->set(dpp, &block, *y) < 0) {
     ldpp_dout(dpp, 0) << "D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::" << __func__ << "(): BlockDirectory set method failed." << dendl;
}

Reply:
How about block.localWeight? Do you use the localWeight that is persisted in the cache, or the block directory localWeight for eviction etc. If yes then the updated localWeight needs to be returned back from the update method and set correctly everywhere.

Reply:
The `localWeight` is stored locally by the policy, not in the directory. The `globalWeight` is stored in the directory and should be transferred over. The `localWeight` is already being updated for those blocks in the `update` method with the new values being stored locally. The `globalWeight` only needs to be incremented for remote accesses.